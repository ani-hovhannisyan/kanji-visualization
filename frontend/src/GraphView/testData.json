{
  "nodes": [
    {
      "id": "山",
      "isMain": true
    },
    {
      "id": "登"
    },
    {
      "id": "士"
    },
    {
      "id": "富"
    },
    {
      "id": "脈"
    },
    {
      "id": "林"
    },
    {
      "id": "頂"
    }
  ],
  "links": [
    {
      "source": "登",
      "target": "山"
    },
    {
      "source": "富",
      "target": "士"
    },
    {
      "source": "士",
      "target": "山"
    },
    {
      "source": "山",
      "target": "脈"
    },
    {
      "source": "山",
      "target": "林"
    },
    {
      "source": "山",
      "target": "頂"
    }
  ],
  "info": [
    {
      "id": "山",
      "isMain": true,
      "onyomi": ["サン", "セン"],
      "kunyomi": ["ヤマ"],
      "meaning": ["mountain"]
    },
    {
      "id": "登",
      "onyomi": ["ト", "トウ"],
      "kunyomi": ["ノボる"],
      "meaning": ["ascend", "climb up"]
    }
  ]
}
