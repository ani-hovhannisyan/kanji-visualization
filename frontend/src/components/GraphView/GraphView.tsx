import React from "react";
import ForceGraph2D, { NodeObject } from "react-force-graph-2d";

interface Props {
  graphData?: GraphMatrix;
  setKanjiInput: React.Dispatch<React.SetStateAction<string>>;
}

const GraphView: React.VFC<Props> = (props) => {
  const handleNodeClick = (node: NodeObject) => {
    const value = node.id?.toString();

    // Validate the input and draw a new result in the graph
    props.setKanjiInput(value || "");
  };

  return (
    <ForceGraph2D
      height={window.innerHeight * 0.6}
      graphData={props.graphData}
      onNodeClick={handleNodeClick}
      linkDirectionalArrowLength={8}
      linkDirectionalArrowRelPos={0.8}
      nodeCanvasObject={(node: ExtendedNodeObject, ctx, globalScale) => {
        if (!node.id) {
          return;
        }

        if (node.isMain === "true") {
          node.x = 0;
          node.y = 0;
        }

        const size = 20 / globalScale;
        const mag = 1;
        const width = size * node.id.toString().length * mag;
        const height = size * mag;
        const x = node.x || 0;
        const y = node.y || 0;

        ctx.fillStyle = node.isMain ? "#2196f3" : "#3d5afe";
        ctx.beginPath();
        ctx.ellipse(x, y, width, height, 0, 0, 2 * Math.PI);
        ctx.closePath();
        ctx.fill();

        const label = node.id as string;
        const fontSize = size;

        ctx.font = `${fontSize}px Sans-Serif`;
        ctx.textAlign = "center";
        ctx.textBaseline = "middle";
        ctx.fillStyle = "white";
        ctx.fillText(label, x, y);
      }}
    />
  );
};

export default GraphView;
