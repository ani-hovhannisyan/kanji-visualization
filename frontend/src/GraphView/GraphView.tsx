import React, { useRef } from "react";
import ForceGraph2D, { NodeObject } from "react-force-graph-2d";
import testData from "./testData.json";

interface ExtendedNodeObject extends NodeObject {
  isMain?: string;
}

interface GraphData {
  nodes?: ExtendedNodeObject[];
  links?: {
    source: string;
    target: string;
  }[];
}

interface Props {
  graphData?: GraphData;
}

const GraphView: React.VFC<Props> = (props) => {
  const ref = useRef();

  const handleNodeClick = (node: NodeObject) => {
    console.log("Node clicked: ", node.id);
  };

  return (
    <ForceGraph2D
      height={400}
      ref={ref}
      graphData={testData}
      onNodeClick={handleNodeClick}
      linkDirectionalArrowLength={8}
      linkDirectionalArrowRelPos={0.8}
      nodeCanvasObject={(node: ExtendedNodeObject, ctx, globalScale) => {
        if (node.isMain === "true") {
          node.x = 0;
          node.y = 0;
        }

        const size = 20 / globalScale;

        ctx.fillStyle = node.isMain ? "#2196f3" : "#3d5afe";
        ctx.beginPath();
        ctx.arc(node.x || 0, node.y || 0, size, 0, 2 * Math.PI);
        ctx.closePath();
        ctx.fill();

        const label = node.id as string;
        const fontSize = size;

        ctx.font = `${fontSize}px Sans-Serif`;
        ctx.textAlign = "center";
        ctx.textBaseline = "middle";
        ctx.fillStyle = "white";
        ctx.fillText(label, node.x || 0, node.y || 0);
      }}
    />
  );
};

export default GraphView;
