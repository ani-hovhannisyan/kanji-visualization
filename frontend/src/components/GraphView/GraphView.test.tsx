import React from "react";
import { render } from "@testing-library/react";
import "jest-canvas-mock";
import GraphView from "./GraphView";
import testData from "./testData.json";

describe("GraphView", () => {
  test("rendered with test data", async () => {
    const mock = jest.fn();

    const { container } = render(
      <GraphView graphData={testData} setKanjiInput={mock} />
    );
    expect(container).toMatchSnapshot();
  });
});
