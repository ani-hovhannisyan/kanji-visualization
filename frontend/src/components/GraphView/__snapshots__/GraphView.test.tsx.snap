// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphView rendered with test data 1`] = `

<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-a8ovzx-MuiPaper-root"
    id="graph-view"
  >
    <h2
      class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      id="graph-view-title"
    >
      Graph
    </h2>
    <div>
      <div
        class="force-graph-container"
        style="position: relative;"
      >
        <canvas
          height="768"
          style="width: 889.6px; height: 768px;"
          width="889"
        />
        <div
          class="graph-tooltip"
        />
      </div>
    </div>
  </div>

//<div class="force-graph-container" >
//      <canvas
//        height="460"
//        style="width: 1024px; height: 460px;"
//        width="1024"
//      />
//      <div class="graph-tooltip" />

</div>
